File: xmv.py
#! /usr/bin/env python3

from json import load, dumps
from os import environ, path
from sys import argv, exit
import subprocess


config_file = path.join(environ["HOME"], ".config/xmv")

config = {
    "patterns": {}
}


def run_zsh(command):
    return subprocess.run(["/usr/bin/zsh", "-ic", command], capture_output=True, text=True)

def run_all_patterns(config):
    successful_patterns = []

    for old_pattern, new_pattern in config["patterns"].items():
        result = run_zsh(f"zmv -n '{old_pattern}' '{new_pattern}'")

        if result.returncode != 0:
            continue

        samples = result.stdout.split("\n")
        successful_patterns.append((
            old_pattern,
            new_pattern,
            samples,
        ))

    with open(config_file, "w") as file:
        file.write(dumps(config))
    file.close()

    for index, pattern in enumerate(successful_patterns):
        old_pattern, new_pattern, samples = pattern
        print(f"{index + 1}. [{old_pattern} -> {new_pattern}] {samples[0]}")

    print("q. quit\n")

    return successful_patterns

def choose_pattern(successful_patterns):
    chosen_pattern_index = input("choose option: ")

    if chosen_pattern_index.lower() in ["q", ""]:
        exit(0)

    if not chosen_pattern_index.isnumeric():
        exit(1)

    chosen_pattern_index = int(chosen_pattern_index)

    if chosen_pattern_index > len(successful_patterns):
        exit(1)

    return successful_patterns[chosen_pattern_index - 1]

def run_one_pattern(pattern):
    old_pattern, new_pattern, samples = pattern

    print("\n".join(samples))

    proceed = input("proceed? y/n ")

    if proceed.lower() != "y":
        exit(0)

    result = run_zsh(f"zmv '{old_pattern}' '{new_pattern}'")

    print(result.stderr)
    print(result.stdout)

    exit(result.returncode)

def main():
    try:
        with open(config_file) as file:
            config = load(file)
        file.close()
    except FileNotFoundError:
        with open(config_file, "w") as file:
            file.write(dumps(config))
        file.close()

    if len(argv) == 3:
        old_pattern = argv[1]
        new_pattern = argv[2]

        config["patterns"][old_pattern] = new_pattern

    successful_patterns = run_all_patterns(config)
    pattern = choose_pattern(successful_patterns)
    run_one_pattern(pattern)

if __name__ == "__main__":
    main()

File: xmv.go
package main

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"sync"
	"time"
)

const (
	ConfigFile        = "/.config/xmv"
	Timeout           = 10 * time.Second
	RunAllPatternsCmd = "zmv -n '%s' '%s'"
	RunOnePatternsCmd = "zmv '%s' '%s'"
)

type Config struct {
	Patterns map[string]string
}

type result struct {
	oldPattern string
	newPattern string
	samples    []string
}

var (
	config = Config{
		Patterns: make(map[string]string),
	}
	configFile = os.Getenv("HOME") + ConfigFile
)

func main() {
	f, err := os.Open(configFile)
	checkError(err, 1)

	err = json.NewDecoder(f).Decode(&config)
	checkError(err, 1)

	if len(os.Args) == 4 {
		oldPattern := os.Args[1]
		newPattern := os.Args[2]
		config.Patterns[oldPattern] = newPattern

		saveConfig()
	}

	succesfulPatterns := runAllPatterns()
	pattern := choosePattern(succesfulPatterns)
	runOnePattern(pattern)
}

func runZSH(command string) (string, int, error) {
	ctx, cancel := context.WithTimeout(context.Background(), Timeout)
	defer cancel()

	exitCode := 0
	zshPath := findZSH()
	cmd := exec.CommandContext(ctx, zshPath, "-ic", command)
	output, err := cmd.Output()

	if exiterr, ok := err.(*exec.ExitError); ok {
		exitCode = exiterr.ExitCode()
	} else if err != nil {
		checkError(err, 1)
	}

	return string(output), exitCode, err
}

func findZSH() string {
	path, err := exec.LookPath("zsh")
	checkError(err, 1)

	return path
}

func runAllPatterns() []result {
	successfulPatterns := make([]result, 0)

	var wg sync.WaitGroup
	ch := make(chan result, 0)

	for oldPattern, newPattern := range config.Patterns {
		wg.Add(1)

		go func(oldPattern, newPattern string) {
			defer wg.Done()
			output, exitCode, _ := runZSH(fmt.Sprintf(RunAllPatternsCmd, oldPattern, newPattern))

			if exitCode != 0 {
				return
			}

			ch <- result{
				oldPattern: oldPattern,
				newPattern: newPattern,
				samples:    strings.Split(output, "\n"),
			}
		}(oldPattern, newPattern)
	}

	go func() {
		for result := range ch {
			successfulPatterns = append(successfulPatterns, result)
		}
	}()

	wg.Wait()

	return successfulPatterns
}

func choosePattern(successfulPatterns []result) result {
	for index, result := range successfulPatterns {
		fmt.Printf("%d. [%s -> %s] %s\n", index+1, result.oldPattern, result.newPattern, result.samples[0])
	}

	v := input("choose option: ")

	if v == "" || strings.ToLower(v) == "q" {
		os.Exit(0)
	}

	index, err := strconv.Atoi(v)
	checkError(err, 1)

	if l := len(successfulPatterns); index > l {
		fmt.Fprintf(os.Stdout, "choose between 1 and %d\n", l)
		os.Exit(1)
	}

	return successfulPatterns[index-1]
}

func runOnePattern(r result) {
	for _, sample := range r.samples {
		fmt.Println(sample)
	}

	proceed := input("proceed? y/n ")

	if strings.ToLower(proceed) != "y" {
		return
	}

	output, exitCode, err := runZSH(fmt.Sprintf(RunOnePatternsCmd, r.oldPattern, r.newPattern))
	checkError(err, exitCode)

	fmt.Println(output)
}

func saveConfig() {
	f, err := os.Create(configFile)
	checkError(err, 1)

	err = json.NewEncoder(f).Encode(config)
	checkError(err, 1)
}

func checkError(err error, exitCode int) {
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(exitCode)
	}
}

func input(label string) string {
	fmt.Print(label)
	reader := bufio.NewReader(os.Stdin)
	v, err := reader.ReadString('\n')
	checkError(err, 1)

	return strings.Trim(v, "\n")
}
